System<T>::is_direct_feedthrough()

amcastro-tri opened the pull request:
Provides a mechanism to flag systems as direct feedthrough.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3101)
<!-- Reviewable:end -->

amcastro-tri commented:
The usual suspects: +@david-german-tri, +@sherm1 

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOkpFvF430_1qA-ThmH:b-t7z0dx)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Nice, this was on my list for today but I hadn't started yet.  @bradking FYI - the implementation of this function should eventually be based on the sparsity matrix, which in turn should be based the functional form.

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/framework/system.h, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkpQSRHT7o1anJdNIm:-KOkpQSRHT7o1anJdNIn:b-i8em8g) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L22)):*
> ```objc
>   virtual ~System() {}
> 
>   /// Evaluates to `true` if system is direct feedthrough and `false` otherwise.
> ```

nit: Possibly `>80`?  Reviewable has been flaky about this recently though so I'm not sure.  Also below.

---

*[drake/systems/framework/system.h, line 23 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkpZToQo0hhwC_rvF-:-KOkpZToQo0hhwC_rvF0:b-h29sky) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L23)):*
> ```objc
> 
>   /// Evaluates to `true` if system is direct feedthrough and `false` otherwise.
>   /// By default systems are assuemed not to be direct feedthrough.
> ```

I think `true` would be a much safer default.

---

*[drake/systems/framework/system.h, line 30 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkpiPFOXP_88AB871G:-KOkpiPFOXP_88AB871H:b-rdiipn) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L30)):*
> ```objc
>   /// one of its inputs. Systems for which outputs only depend on the state of
>   /// the system are not direct feedthrough.
>   /// Direct feedthrough systems define an implicit dependence of their outputs
> ```

I'm not sure this sentence is helpful.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOkpv1nnUahRZn2d6gR:b-3666c6)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/framework/primitives/test/integrator_test.cc, line 90 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkqVOii3pzEs-s282m:-KOkqVOii3pzEs-s282n:b-27x3nr) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/primitives/test/integrator_test.cc#L90)):*
> ```cpp
> 
> // Asserts integrators are not direct feedthrough systems.
> TEST_F(IntegratorTest, IntegratorIsDirectFeedthrough) {
> ```

`IntegratorIsNotDirectFeedthrough`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/framework/system.h, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkpQSRHT7o1anJdNIm:-KOksx9x0ZqddcAD-U0d:bm27c5r) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L22)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> nit: Possibly `>80`?  Reviewable has been flaky about this recently though so I'm not sure.  Also below.
</details>
The `.` is the 80-th character. I cpplinted these files before pushing and reported zero errors.

---

*[drake/systems/framework/system.h, line 23 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkpZToQo0hhwC_rvF-:-KOktD7rCJw-N83nzbDk:bqatvfi) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L23)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> I think `true` would be a much safer default.
</details>
Humm... good point. I was thinking of newbies implementing their particular dynamical systems (typically in the form `xdot = flow(t,x,u)` and that not even know what a direct feedthrough is. I guess that if they forget to implement this method they'll have a hit in performance rather than a failure?

---

*[drake/systems/framework/system.h, line 30 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkpiPFOXP_88AB871G:-KOktm5IGZlZP5upAz9A:b-ytiac2) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L30)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> I'm not sure this sentence is helpful.
</details>
I was trying to add some helpful information for new users of systems. Actually based on a question by @liangfok. @liangfok, is this helpful to you? should I rephrase it differently? 

---

*[drake/systems/framework/primitives/test/integrator_test.cc, line 90 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkqVOii3pzEs-s282m:-KOkuuiLsYfh7cP_Je1T:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/primitives/test/integrator_test.cc#L90)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> `IntegratorIsNotDirectFeedthrough`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
"Direct feedthrough" shouldn't be viewed as a property of a System as a whole. It is a relationship between a particular input port and a particular output port. If I understand correctly, Simulink does this part way by considering "direct feedthrough" to be a property of an input port. I don't object to the idea of having some convenient abbreviations, but I would like us to think about the dependency matrix mapping each input port to each output port. I agree with David that a good default is "every output port depends directly on every input port" but I think we will be able to improve on that later based either on functional form analysis or just on specified dependencies used for caching. 

BTW, Reviewable is killing me with timeouts!

---

Review status: 3 of 4 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOktxEKiAqx5Tor_07i:b4kc7do)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
I agree with everything @sherm1 said above (especially the timeouts), but I think this PR is a fine place to start (except for the default, see below).

---

Review status: 3 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/framework/system.h, line 23 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkpZToQo0hhwC_rvF-:-KOkvcImSSGrnedP1Te_:b-olrmsd) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L23)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Humm... good point. I was thinking of newbies implementing their particular dynamical systems (typically in the form `xdot = flow(t,x,u)` and that not even know what a direct feedthrough is. I guess that if they forget to implement this method they'll have a hit in performance rather than a failure?
</details>
I can't comment on the solver implications, but as for diagram wiring: 

If we default to `false` and the system is actually direct-feedthrough, we are likely to compute the output using stale input.  

If we default to `true` and the system is not actually direct-feedthrough, there will be no adverse effects unless there is an algebraic loop involving the system, in which case we will throw an exception.  

The latter seems like a much better outcome.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOkxuqKgFQ_j_vBXDqa:bexhpt6)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
I'm just arguing for the right abstraction, not necessarily for a functional change here. But the comment should not say "this system is direct feedthrough" but rather something like "the input port value feeds through directly to the output port". What do you think about using "dependency" as the terminology rather than feed-through? For example, "the current value of the output port depends on the current value of the input port".

---

Review status: 3 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/framework/system.h, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkpQSRHT7o1anJdNIm:-KOkzSO23A7aZLh2ejQG:b1bars5) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L22)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> The `.` is the 80-th character. I cpplinted these files before pushing and reported zero errors.
</details>
BTW, Reviewable is screwing up the line wrapping. I filed [an issue](https://github.com/Reviewable/Reviewable/issues/385) there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOkzjlr2F1tqS_gxw3m:b-4avoed)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I like the idea of flagging inputs as direct feedthrough or not, I think that's the right way to do it in general. Not to be annoying but, with #2955 we will be able to flag each input port separately :-). Like `InputPort input("my input"); input_port.flag_as_direct_feedthrough()`.

I agree with @david-german-tri this being a good start also. Once we have the capability to flag input ports as direct feedthrough (or not) the method `System<T>::is_direct_feedthrough` will scan its ports (and subsystems for `Diagram`'s) to determine if the system as a whole is direct feedthrough. 

---

Review status: 3 of 4 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOkze9Jhjxe5uMB4lI9:b7eib74)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 2 of 4 files at r1.
Review status: 3 of 4 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/framework/system.h, line 23 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkpZToQo0hhwC_rvF-:-KOl-xOZX5iDnsZ9AATg:b-eydv9o) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L23)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> I can't comment on the solver implications, but as for diagram wiring: 
> 
> If we default to `false` and the system is actually direct-feedthrough, we are likely to compute the output using stale input.  
> 
> If we default to `true` and the system is not actually direct-feedthrough, there will be no adverse effects unless there is an algebraic loop involving the system, in which case we will throw an exception.  
> 
> The latter seems like a much better outcome.
</details>
BTW, agreed! It isn't a performance issue but whether an apparent loop in the system diagram is an actual simultaneous mutual dependency. We are forbidding those in Drake; in Simulink that would cause a nonlinear solver to be invoked at every time step and would be both a performance and robustness problem.

---

*[drake/systems/framework/system.h, line 32 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOl0lSysgoQLBGVS9RQ:-KOl0lSysgoQLBGVS9RR:bhpa8fx) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L32)):*
> ```objc
>   /// Direct feedthrough systems define an implicit dependence of their outputs
>   /// on their inputs and therefore specific solving strateges must be applied.
>   virtual bool is_direct_feedthrough() const { return false;}
> ```

Actually I do think this needs a change. If we can't say that direct feedthrough is a a System property, we could say a System `has_any_direct_feedthrough()` but not `is_...`. There should also be a TODO saying that this API is likely to change to provide more resolution.

---

*[drake/systems/framework/primitives/adder.h, line 26 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOl23qEjFWhXCk21s4B:-KOl23qEjFWhXCk21s4C:bfbf93j) ([raw file](https://github.com/robotlocomotion/drake/blob/6b544e289c36a5350633de1f2da541dd04c8a2ac/drake/systems/framework/primitives/adder.h#L26)):*
> ```objc
>   /// Flags adders as direct feedthrough systems since their outputs directly
>   /// depend on their inputs by a simple, state independent, algebraic relation.
>   bool is_direct_feedthrough() const override { return true;}
> ```

This can use the default implementation when it is `true`.

---

*[drake/systems/framework/primitives/test/integrator_test.cc, line 91 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOl2Ihafmrct5FBY70E:-KOl2Ihafmrct5FBY70F:b-f0gmef) ([raw file](https://github.com/robotlocomotion/drake/blob/6b544e289c36a5350633de1f2da541dd04c8a2ac/drake/systems/framework/primitives/test/integrator_test.cc#L91)):*
> ```cpp
> // Asserts integrators are not direct feedthrough systems.
> TEST_F(IntegratorTest, IntegratorIsNotDirectFeedthrough) {
>   EXPECT_FALSE(integrator_->is_direct_feedthrough());
> ```

Integrator will need to implement the `has_any_direct_feedthrough() = false` method once the default has flipped.

---

*[drake/systems/framework/primitives/test/integrator_test.cc, line 91 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOl2YJ4D33jHcv1d58K:-KOl2YJ4D33jHcv1d58L:b-y8w3jx) ([raw file](https://github.com/robotlocomotion/drake/blob/6b544e289c36a5350633de1f2da541dd04c8a2ac/drake/systems/framework/primitives/test/integrator_test.cc#L91)):*
> ```cpp
> // Asserts integrators are not direct feedthrough systems.
> TEST_F(IntegratorTest, IntegratorIsNotDirectFeedthrough) {
>   EXPECT_FALSE(integrator_->is_direct_feedthrough());
> ```

BTW, the "initial condition" InputPort to the integrator likely does feed through directly to its OutputPort during initial condition computations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
BTW, @amcastro-tri I'm not saying we should flag inputs like Simulink (assuming I even understood correctly what Simulink is doing). I'm saying direct feedthrough is like a "sparsity" property relating an (output_port,input_port) pair. Given that dependency matrix, you could then answer these questions:
- does the System have any direct feedthrough at all? (Dependency(out_i,in_j)=true for some (i,j))
- does in_port_x ever feed through directly to an output port? (Dependency(out_i,in_port_x)=true for some i)
- does out_port_y ever depend directly on an input port? (Dependency(out_port_y,in_j)=true for some j)

So marking in_port_x as "direct feedthrough" should be an abbreviation for marking every (out_i, in_port_x) pair as true in the dependency matrix.

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOl5_9KzvqO8s4FMBxg:b-8akiir)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
P.S. We need almost exactly this information for correct caching -- it would be nice if we could collect it just once.

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOl7WkjoGKyL2SWdCsJ:bh776b0)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/framework/system.h, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkpQSRHT7o1anJdNIm:-KOovqAlkoZNapBxTAiS:b-6wk9yd) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L22)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, Reviewable is screwing up the line wrapping. I filed [an issue](https://github.com/Reviewable/Reviewable/issues/385) there.
</details>
And it was fixed! nice!

---

*[drake/systems/framework/system.h, line 23 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkpZToQo0hhwC_rvF-:-KOp-SubVF4kUqN61dNq:bmu2bl5) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L23)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, agreed! It isn't a performance issue but whether an apparent loop in the system diagram is an actual simultaneous mutual dependency. We are forbidding those in Drake; in Simulink that would cause a nonlinear solver to be invoked at every time step and would be both a performance and robustness problem.
</details>
I default to systems being direct feedthrough now.

---

*[drake/systems/framework/system.h, line 32 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOl0lSysgoQLBGVS9RQ:-KOp259VY78o_x247r-A:b-119r21) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L32)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Actually I do think this needs a change. If we can't say that direct feedthrough is a a System property, we could say a System `has_any_direct_feedthrough()` but not `is_...`. There should also be a TODO saying that this API is likely to change to provide more resolution.
</details>
Agree. I refactored `is_direct_feedthrough` to `has_any_direct_feedthrough`

---

*[drake/systems/framework/primitives/adder.h, line 26 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOl23qEjFWhXCk21s4B:-KOp2GEGvDE4VjFG5dOV:b6l8q99) ([raw file](https://github.com/robotlocomotion/drake/blob/6b544e289c36a5350633de1f2da541dd04c8a2ac/drake/systems/framework/primitives/adder.h#L26)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This can use the default implementation when it is `true`.
</details>
Correct. I decided however to provide an implementation so it would appear explicitly in the documentation. 

---

*[drake/systems/framework/primitives/test/integrator_test.cc, line 91 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOl2Ihafmrct5FBY70E:-KOp2UfLjnVVzqi66TCs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6b544e289c36a5350633de1f2da541dd04c8a2ac/drake/systems/framework/primitives/test/integrator_test.cc#L91)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Integrator will need to implement the `has_any_direct_feedthrough() = false` method once the default has flipped.
</details>
Done.

---

*[drake/systems/framework/primitives/test/integrator_test.cc, line 91 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOl2YJ4D33jHcv1d58K:-KOp2YFvOapB3W6x_tMD:b-95x280) ([raw file](https://github.com/robotlocomotion/drake/blob/6b544e289c36a5350633de1f2da541dd04c8a2ac/drake/systems/framework/primitives/test/integrator_test.cc#L91)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, the "initial condition" InputPort to the integrator likely does feed through directly to its OutputPort during initial condition computations.
</details>
Good observation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/framework/system.h, line 30 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOkpiPFOXP_88AB871G:-KOp2peIhIvyNULKRNKV:bq2t88w) ([raw file](https://github.com/robotlocomotion/drake/blob/1e006a9995bb4e408bb02096eb6e2735dc301ac1/drake/systems/framework/system.h#L30)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I was trying to add some helpful information for new users of systems. Actually based on a question by @liangfok. @liangfok, is this helpful to you? should I rephrase it differently? 
</details>
On a second thought I agree. Removed and updated with the new documentation. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed -- proposed some hopefully-clarifying comments.

---

Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/framework/system.h, line 27 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOpF3Gqg1kdQGgU39KU:-KOpF3Gqg1kdQGgU39KV:b7jvhxy) ([raw file](https://github.com/robotlocomotion/drake/blob/d547caca817580ba80dc3a1178b70329285802c7/drake/systems/framework/system.h#L27)):*
Suggest rewording:
> By default we assume that there is direct feedthrough of values from
> every input port to every output port. This is a conservative assumption
> that ensures we detect and can prevent formation of algebraic loops (implicit
> computations) in system diagrams. Any System for which *none* of the
> input ports *ever* feeds through to *any* of the output ports should override
> this method to return `false`.

Also add "TODO: need to provide a more nuanced approach that can represent pairwise (input,output) feed through. For example, we should be able to express that initial conditions feed through an integrator but the dynamic signal during simulation does not."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOpJFcGcv0PDIyJTFTR:bb4c0br)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 4 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/framework/system.h, line 27 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOpF3Gqg1kdQGgU39KU:-KOq0AvyOT8bYudZVPXc:b-5szkb4) ([raw file](https://github.com/robotlocomotion/drake/blob/d547caca817580ba80dc3a1178b70329285802c7/drake/systems/framework/system.h#L27)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Suggest rewording:
> > By default we assume that there is direct feedthrough of values from
> > every input port to every output port. This is a conservative assumption
> > that ensures we detect and can prevent formation of algebraic loops (implicit
> > computations) in system diagrams. Any System for which *none* of the
> > input ports *ever* feeds through to *any* of the output ports should override
> > this method to return `false`.
> 
> Also add "TODO: need to provide a more nuanced approach that can represent pairwise (input,output) feed through. For example, we should be able to express that initial conditions feed through an integrator but the dynamic signal during simulation does not."
</details>
Accepted your suggestion and added TODO. I also added TODO in `integrator.h` regarding the ability to express that initial conditions do feed through but not its input signal.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOqGY_w-jXkmh21HPz9:bnfp4nl)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 6 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[drake/systems/framework/primitives/adder.h, line 26 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOl23qEjFWhXCk21s4B:-KOqI6e_1vlhmO-mocXh:bonhngr) ([raw file](https://github.com/robotlocomotion/drake/blob/6b544e289c36a5350633de1f2da541dd04c8a2ac/drake/systems/framework/primitives/adder.h#L26)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Correct. I decided however to provide an implementation so it would appear explicitly in the documentation. 
</details>
BTW I agree with @amcastro-tri, this is worth doing inside `systems/framework`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOqIJo0hwvBlOljpXYx:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Did a quick pass through this PR and left a few comments.

---

Reviewed 4 of 6 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 101 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOqI0-RivNj5uNz_mfp:-KOqI0-SNNsn7W-AcCs2:bwscwr1) ([raw file](https://github.com/robotlocomotion/drake/blob/6659feb45155ac2cf430b6838827361efd708dac/drake/examples/spring_mass/spring_mass_system.h#L101)):*
> ```objc
>   using MyOutput = systems::SystemOutput<double>;
> 
>   /// The input force to this system is not directly fed through to the output.
> ```

In my mind, the word "directly" implies that the input force to the system could, theoretically, _indirectly_ be fed through to the output. Is this implication correct?

---

*[drake/systems/framework/system.h, line 27 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOqHkLtJIal9F1cjmhU:-KOqHkLtJIal9F1cjmhV:b-qe9bfu) ([raw file](https://github.com/robotlocomotion/drake/blob/6659feb45155ac2cf430b6838827361efd708dac/drake/systems/framework/system.h#L27)):*
> ```objc
>   /// By default we assume that there is direct feedthrough of values from
>   /// every input port to every output port. This is a conservative assumption
>   /// that ensures we detect and can prevent formation of algebraic loops
> ```

Missing "the" after "prevent".

Alternatively, delete "formation of" for conciseness and clarity.

---

*[drake/systems/framework/primitives/test/adder_test.cc, line 88 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOqHRL-LNfbDfFg4rd1:-KOqHRL-LNfbDfFg4rd2:b11r7t0) ([raw file](https://github.com/robotlocomotion/drake/blob/6659feb45155ac2cf430b6838827361efd708dac/drake/systems/framework/primitives/test/adder_test.cc#L88)):*
> ```cpp
> }
> 
> // Asserts adders are direct feedthrough systems.
> ```

BTW: consider adding "that" after "Asserts".

---

*[drake/systems/framework/primitives/test/integrator_test.cc, line 89 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOqH9havOIouNwqkRUF:-KOqH9hbuITI-ttDSJla:b2dz0uf) ([raw file](https://github.com/robotlocomotion/drake/blob/6659feb45155ac2cf430b6838827361efd708dac/drake/systems/framework/primitives/test/integrator_test.cc#L89)):*
> ```cpp
> }
> 
> // Asserts integrators are not direct feedthrough systems.
> ```

BTW: consider adding "that" after "Asserts"

Also consider adding "since the output depends on both the current input and past input" after "systems".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOqIBJHMAJVYkrLPBbZ:b-c0d9hr)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 101 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOqI0-RivNj5uNz_mfp:-KOqZ4g5pFJLb13yBdtp:b-w7ppwt) ([raw file](https://github.com/robotlocomotion/drake/blob/6659feb45155ac2cf430b6838827361efd708dac/drake/examples/spring_mass/spring_mass_system.h#L101)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> In my mind, the word "directly" implies that the input force to the system could, theoretically, _indirectly_ be fed through to the output. Is this implication correct?
</details>
Right. For example, you might integrate the input and report the integral as an output. That would be indirect feedthrough. Another example would be if the input were delayed and then output; that is also indirect feedthrough so can't produce algebraic loops.

---

*[drake/systems/framework/system.h, line 27 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOqHkLtJIal9F1cjmhU:-KOqZYcBnB09NKsDmSRW:b-45n3sk) ([raw file](https://github.com/robotlocomotion/drake/blob/6659feb45155ac2cf430b6838827361efd708dac/drake/systems/framework/system.h#L27)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Missing "the" after "prevent".
> 
> Alternatively, delete "formation of" for conciseness and clarity.
</details>
BTW, I think it reads fine as written although I don't object to and extra "the". It would not be correct to remove "formation of" because it is the *formation* of those loops that we want to prevent (we're making the wiring diagram).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
@liangfok , I addressed your comments in this last revision. PTAL!

---

Review status: 3 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 101 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOqI0-RivNj5uNz_mfp:-KOtrWWGrG3d8vk76qy-:b-jvc479) ([raw file](https://github.com/robotlocomotion/drake/blob/6659feb45155ac2cf430b6838827361efd708dac/drake/examples/spring_mass/spring_mass_system.h#L101)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Right. For example, you might integrate the input and report the integral as an output. That would be indirect feedthrough. Another example would be if the input were delayed and then output; that is also indirect feedthrough so can't produce algebraic loops.
</details>
Does @sherm1's reply answer your question @liangfok ?

---

*[drake/systems/framework/system.h, line 27 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOqHkLtJIal9F1cjmhU:-KOtwZV2H60iDFokszNo:bra0wd0) ([raw file](https://github.com/robotlocomotion/drake/blob/6659feb45155ac2cf430b6838827361efd708dac/drake/systems/framework/system.h#L27)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, I think it reads fine as written although I don't object to and extra "the". It would not be correct to remove "formation of" because it is the *formation* of those loops that we want to prevent (we're making the wiring diagram).
</details>
added suggested "the" before "prevent".

---

*[drake/systems/framework/primitives/test/adder_test.cc, line 88 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOqHRL-LNfbDfFg4rd1:-KOtxBMURLa_2yjaHDIv:b-dt2r6p) ([raw file](https://github.com/robotlocomotion/drake/blob/6659feb45155ac2cf430b6838827361efd708dac/drake/systems/framework/primitives/test/adder_test.cc#L88)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: consider adding "that" after "Asserts".
</details>
considered "that" and reworded to reflect discussions in this PR.

---

*[drake/systems/framework/primitives/test/integrator_test.cc, line 89 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOqH9havOIouNwqkRUF:-KOtxWKbgH_31Mb-JjUf:btkrgxi) ([raw file](https://github.com/robotlocomotion/drake/blob/6659feb45155ac2cf430b6838827361efd708dac/drake/systems/framework/primitives/test/integrator_test.cc#L89)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: consider adding "that" after "Asserts"
> 
> Also consider adding "since the output depends on both the current input and past input" after "systems".
</details>
Considered "that" and and reworded to reflect discussions in this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOtxbJXOwxtSJOLniD-:b-sdnfc7)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 101 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOqI0-RivNj5uNz_mfp:-KOtzQsZase-yigUdAc1:bbiehrt) ([raw file](https://github.com/robotlocomotion/drake/blob/6659feb45155ac2cf430b6838827361efd708dac/drake/examples/spring_mass/spring_mass_system.h#L101)):*
Yes, @sherm1's reply indicates that indirect feedthrough is possible. For consistency and to avoid future questions, I recommend changing this comment to be:

> The input force to this system is not direct feedthrough.

Notice that that I use the phrase "direct feedthrough" instead of "directly fed through" since it appears that "direct feedthrough" is a well known and understood term in the Simulink community.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Just one comment (marked myself as satisfied).

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-:-KOu-15gWf3y7ktFvqlo:b-h6lzob)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 5 of 6 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 101 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOqI0-RivNj5uNz_mfp:-KOu7NDErbYiKw0KU8En:b-hu8krw) ([raw file](https://github.com/robotlocomotion/drake/blob/6659feb45155ac2cf430b6838827361efd708dac/drake/examples/spring_mass/spring_mass_system.h#L101)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Yes, @sherm1's reply indicates that indirect feedthrough is possible. For consistency and to avoid future questions, I recommend changing this comment to be:
> 
> > The input force to this system is not direct feedthrough.
> 
> Notice that that I use the phrase "direct feedthrough" instead of "directly fed through" since it appears that "direct feedthrough" is a well known and understood term in the Simulink community.
</details>
Interesting, I wrote this doxydoc this way thinking about you @liangfok. But, changed as you suggested if that makes it clearer now :-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 101 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOqI0-RivNj5uNz_mfp:-KOu9FsGutQT4L0_FyUb:bsfhzqn) ([raw file](https://github.com/robotlocomotion/drake/blob/6659feb45155ac2cf430b6838827361efd708dac/drake/examples/spring_mass/spring_mass_system.h#L101)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Interesting, I wrote this doxydoc this way thinking about you @liangfok. But, changed as you suggested if that makes it clearer now :-)
</details>
Sorry, I'm still slowly getting immersed in the Simulink culture :smile: 

---

*[drake/examples/spring_mass/spring_mass_system.h, line 101 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3101#-KOu9JCvdmvsDzxNAs9n:-KOu9JCvdmvsDzxNAs9o:b9qioqz) ([raw file](https://github.com/robotlocomotion/drake/blob/d8f779974ae66484da55febaa74d8ee29410297b/drake/examples/spring_mass/spring_mass_system.h#L101)):*
> ```objc
>   using MyOutput = systems::SystemOutput<double>;
> 
>   /// The input force to this system is not direct feedthrough.
> ```

BTW, should this be "feedthrough to the output"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3101)*
<!-- Sent from Reviewable.io -->
